### PR TITLE
add dcgan test

### PR DESCRIPTION
**Pull request by houseroad**
This commit actually does three things:
1) Add support for ConvTranspose op in torch/csrc/autograd/functions/toffee/convolution.cpp. So we can support ConvTranspose2d which is used by the dcgan model.
2) Fix the dim primspec for Squeeze op.
3) Add dcgan pytorch-caffe2 test.

### CODE REVIEW

File: test/model_defs/caffe2_pytorch_test_models.py [hunk 1]
```diff
@@ -44,12 +44,13 @@
 
 
 class TestCaffe2Backend(unittest.TestCase):
-    def run_model_test(self, model, train, batch_size):
+    def run_model_test(self, model, train, batch_size, x=None):
```
Comment (ezyang) [inline test/model_defs/caffe2_pytorch_test_models.py]: Would be far better to have a descriptive name for this parameter. Input?

File: torch/autograd/_functions/tensor.py [hunk 2]
```diff
@@ -380,7 +380,15 @@ class Squeeze(InplaceFunction):
     def primspec(input, dim, inplace=False):
         if inplace:
             return None
-        return torch.toffee.op("Squeeze", input, dims=[dim])
+        if dim is None:
+            dim = []
+            for i, size in enumerate(input.size()):
```
Comment (ezyang) [inline torch/autograd/_functions/tensor.py:385]: Uh... did we land something in dirty_jit that changed the ABI of the passed in inputs? If not, I am not sure how this works at all, because input is going to be an integer. In any case, I don't see an explicit test for this case.

File: torch/csrc/autograd/functions/toffee/convolution.cpp [hunk 3]
```diff
@@ -23,45 +29,91 @@ void ConvForward::primspec(PrimSpecContext* ctx, jit::node_list inputs, jit::nod
...
-  attr->set_name("kernels");
-  for (int kernel : kernel_size) {
-    attr->add_ints(kernel);
+  if (transposed == false) {
+    attr->set_name("kernels");
+    for (int kernel : kernel_size) {
+      attr->add_ints(kernel);
+    }
+  } else {
+    attr->set_name("kernel");
```
Comment (ezyang) [inline torch/csrc/autograd/functions/toffee/convolution.cpp]: Is it actually true that per-dimension kernel settings are not supported in the transposed version?

File: torch/csrc/autograd/functions/toffee/convolution.cpp [hunk 4]
```diff
@@ -23,45 +29,91 @@ void ConvForward::primspec(PrimSpecContext* ctx, jit::node_list inputs, jit::nod
...
+
+  // attribute dilation(s).
+  if (transposed == false) {
+    attr = p_n->add_attribute();
+    attr->set_name("dilations");
+    for (int d : dilation) {
+      attr->add_ints(d);
+    }
+  } else {
+    // ConvTranspose in Caffe2 only supports dilation=1.
```
Comment (ezyang) [inline torch/csrc/autograd/functions/toffee/convolution.cpp:103]: If this is the case for everything else, would be good to have comments accordingly. Actually, I would seriously consider just having one giant if-statement on transposed, since it seems to require you to do completely different things.

**Comment by ezyang** [comment]
I am not merging this commit because I don't understand how the Squeeze primspec could be working.

**Comment by ezyang** [comment]
Perhaps this is a post https://github.com/ezyang/pytorch/pull/112 patch?

File: test/model_defs/caffe2_pytorch_test_models.py (hunk 1 above)
Comment (houseroad) [reply test/model_defs/caffe2_pytorch_test_models.py]: I will change x to input.

File: torch/autograd/_functions/tensor.py (hunk 2 above)
Comment (houseroad) [reply torch/autograd/_functions/tensor.py:385]: I think in this case, input is a tensor. dims should be ints instead i, so it has to be passed as a tuple, otherwise, we cannot pass the check. Caffe2 doesn't support no dims, we have to explicitly indicate which dim we want to squeeze.

File: torch/csrc/autograd/functions/toffee/convolution.cpp (hunk 3 above)
Comment (houseroad) [reply torch/csrc/autograd/functions/toffee/convolution.cpp]: Yes, it's true. I have confirmed this with Junjie.

File: torch/csrc/autograd/functions/toffee/convolution.cpp (hunk 4 above)
Comment (houseroad) [reply torch/csrc/autograd/functions/toffee/convolution.cpp:103]: Yep, I also thought about this. But I find we can share some code for two different cases, so I choose use small if branch to set the attributes for two ops. I will add more comments for other attributes.

**Comment by houseroad** [comment]
No, it's not a post #112 patch... We need to figure out how to merge it.
